components/raftstore-v2/src/operation/life.rs

```diff
@@ -0,0 +1,281 @@
+            ctx.router.send(region_id, PeerMsg::RaftMessage(msg))
+        {
+            m
+        } else {
+            return;
+        };
+        let msg_type = msg.get_message().get_msg_type();
+        let from_peer = msg.get_from_peer();
+        let to_peer = msg.get_to_peer();
+        // Now the peer should not exist.
+        debug!(
+            self.logger(),
+            "handle raft message";
+            "from_peer_id" => from_peer.id,
+            "to_peer_id" => to_peer.id,
+            "region_id" => region_id,
+            "msg_type" => %util::MsgType(&msg)
+        );
+        if to_peer.store_id != self.store_id() {
+            ctx.raft_metrics.message_dropped.mismatch_store_id.inc();
+            return;
+        }
+        if !msg.has_region_epoch() {
+            ctx.raft_metrics.message_dropped.mismatch_region_epoch.inc();
+            return;
+        }
+        // TODO: maybe we need to ack the message to confirm the peer is destroyed.
+        if msg.get_is_tombstone() || msg.has_merge_target() {
+            // Target tombstone peer doesn't exist, so ignore it.
+            ctx.raft_metrics.message_dropped.stale_msg.inc();
+            return;
+        }
+        let from_epoch = msg.get_region_epoch();
+        let local_state = match ctx.engine.get_region_state(region_id) {
+            Ok(s) => s,
+            Err(e) => {
+                error!(self.logger(), "failed to get region state"; "region_id" => region_id, "err" => ?e);
+                return;
+            }
+        };
+        if let Some(local_state) = local_state {
+            // Split will not create peer in v2, so the state must be Tombstone.
+            if local_state.get_state() != PeerState::Tombstone {
+                panic!(
+                    "[region {}] {} peer doesn't exist but has valid local state {:?}",
+                    region_id, to_peer.id, local_state
+                );
+            }
+            // Compared to v1, we rely on leader to confirm destroy actively, so here
+            // skip handling gc for simplicity.
+            let local_epoch = local_state.get_region().get_region_epoch();
+            // The region in this peer is already destroyed
+            if util::is_epoch_stale(from_epoch, local_epoch) {
+                ctx.raft_metrics.message_dropped.region_tombstone_peer.inc();
+                return;
+            }
+            if let Some(local_peer) = util::find_peer(local_state.get_region(), self.store_id()) {
+                if to_peer.id <= local_peer.get_id() {
+                    ctx.raft_metrics.message_dropped.region_tombstone_peer.inc();
+                    return;
+                }
+            }
+        }
+
+        // So the peer must need to be created. We don't need to synchronous with split
+        // as split won't create peer in v2. And we don't check for range
+        // conflict as v2 depends on tablet, which allows conflict ranges.
```

Comment:
Is TiFlash compatible with this behavior?

Reply:
Not yet.

Reply:
No, unless they refactor to use v2.